ci/ci/ci.py

```diff
@@ -121,13 +126,54 @@
 
     batch_client = request.app['batch_client']
     batches = batch_client.list_batches(
-        f'test=1 pr={pr_number}')
+        f'test=1 pr={pr.number}')
     batches = sorted([b async for b in batches], key=lambda b: b.id, reverse=True)
     page_context['history'] = [await b.status() for b in batches]
 
     return await render_template('ci', request, userdata, 'pr.html', page_context)
 
 
+async def retry_pr(wb, pr, request):
+    app = request.app
+    session = await aiohttp_session.get_session(request)
+
+    if pr.batch is None:
+        log.info('retry cannot be requested for PR #{pr.number} because it has no batch')
+        set_message(
+            session,
+            f'Retry cannot be requested for PR #{pr.number} because it has no batch.',
+            'error')
+        raise web.HTTPFound(
```

Comment:
Minor: I try to use return for normal responses, and raise for exceptional/error responses.  That's a bit vague for HTTP status codes, I think elsewhere I've used return for redirects.  You also split the response half in this function and half outside.  I'd return put both redirect returns in this function, and then just return the result of of the shield in the caller.

Reply:
Addressed and switch other users of `raise web.HTTPFound` to `return`.